src/compiler/moduleSpecifiers.ts

```diff
@@ -768,28 +769,22 @@ namespace ts.moduleSpecifiers {
                     }
                 }
                 // If the file is the main module, it can be imported by the package name
-                const mainFileRelative = packageJsonContent.typings || packageJsonContent.types || packageJsonContent.main;
+                const mainFileRelative = packageJsonContent.typings || packageJsonContent.types || packageJsonContent.main || "index.js";
                 if (isString(mainFileRelative)) {
                     const mainExportFile = toPath(mainFileRelative, packageRootPath, getCanonicalFileName);
                     if (removeFileExtension(mainExportFile) === removeFileExtension(getCanonicalFileName(moduleFileToTry))) {
                         return { packageRootPath, moduleFileToTry };
                     }
                 }
             }
-            return { moduleFileToTry };
-        }
-
-        function getExtensionlessFileName(path: string): string {

```

Comment:
This function failed to respect ending preferences as well as the fact that `.mjs`/`.cjs` can never be resolved without extensions. `removeExtensionAndIndexPostFix` handles those things but not checking whether another file prevents dropping the `/index` postfix, so I deleted this function and moved that check into the other. However, I didn’t enable the check for all callers since we’re a little close to the end of the release cycle and I wasn’t trying to make sweeping changes.

Reply:
> as well as the fact that `.mjs`/`.cjs` can never be resolved without extensions.

do we need tests for that case?

Reply:
👍 Done